typing/typecore.ml

```diff
@@ -5871,6 +5960,48 @@ and type_function_cases_expect
     cases, partial, ty_fun
   end
 
+and type_effect_cases
+    : type k . k pattern_category ->
+           _ -> _ -> _ -> Parsetree.case list -> _
+             ->
+           k case list
+  = fun category env ty_res_explained loc caselist conts ->
+  let { ty = ty_res; explanation = _ } = ty_res_explained in
+  let _ = newvar () in
+  (* remember original level *)
+  with_local_level begin fun () ->
+  (* Create a fake abstract type declaration for effect type. *)
+  let decl = {
+    type_params = [];
+    type_arity = 0;
+    type_kind = Type_abstract Definition;
+    type_private = Public;
+    type_manifest = None;
+    type_variance = [];
+    type_separability = [];
+    type_is_newtype = true;
+    type_expansion_scope = Btype.lowest_level;
+    type_loc = loc;
+    type_attributes = [];
+    type_immediate = Unknown;
+    type_unboxed_default = false;
+    type_uid = Uid.mk ~current_unit:(Env.get_unit_name ());
+  }
+  in
+  let name = Ctype.get_new_abstract_name env "effect" in
```

Comment:
Since `effect` is an operator name, type error messages will render this type using the raw literal syntax (`\#effect`)  which looks non-optimal. Since this is a compiler produced name that should only appear in error messages, I propose to use a recognizable special name. Maybe `%eff` using a prefix like for existential type `$a`?